### PR TITLE
[node_modules] Hoisting memory efficiency improvement

### PR DESCRIPTION
**Pull request by larixer**
**What's the problem this PR addresses?**

This PR fixes issue: #1311 

**How did you fix it?**

Current algorithm in `master` does not take into account depth of the hoisted node and because of that it performs a lot of unnecessary work.

This implementation radically reduces memory consumption by hoisting shallowest nodes first. 

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I have verified that all automated PR checks pass.

### CODE REVIEW

**Comment by larixer** [comment]
I have new ideas, how make this algo much faster, though it is already an improvement from master, I'd like to work a bit more on it to make it speedier too.

**Comment by bertho-zero** [comment]
@larixer The link time has increased considerably since the last modifications, I went from 1.3-1.7s to 4.8-5s on https://github.com/bertho-zero/yarn-bug-repro.

**Comment by larixer** [comment]
@bertho-zero I haven't finished yet, I'm transforming the algorithm step by step. I will let you know when I finally arrive to speedier version.

**Comment by larixer** [comment]
@bertho-zero The hoisting should be much faster with latest changes pushed into this PR, I think its close to maximum achievable speed now.

**Comment by bertho-zero** [comment]
@larixer Perfect !

Minimal repro: **1.3s**
Real project: **19s**

**Comment by larixer** [comment]
@bertho-zero I need repro to take a look. 19s seems much, do you refer to link step time, or do you measure hoist time itself via `NM_DEBUG_LEVEL=0 yarn` ?

**Comment by bertho-zero** [comment]
It is a monorepo which contains several GB of node_modules, with master I am at 30s. I am referring to the link step time.

I am trying to make a reproduction, early next week.

**Comment by larixer** [comment]
@bertho-zero Thanks. This PR is about hoisting, Link step time is not only hoisting, its many other things too. To make Link step time more or less the same to v1, you might want to use
`enableGlobalCache: true` and `compressionLevel:0` options in your `.yarnrc.yml`. v1 used uncompressed cache, v2 uses compressed cache by default.

File: packages/yarnpkg-pnpify/sources/hoist.ts [hunk 1]
```diff
@@ -115,6 +113,102 @@ const getHoistedDependencies = (rootNode: HoisterWorkTree): Map<PackageName, Hoi
...
+
+/**
+ * Decouples graph path all the way from decoupled graph node (tree node) up to the node in concern.
+ *
+ * @param nodePath graph path that starts from decoupled graph node up to the node in concern
+ * @returns decoupled node, the last one in `nodePath`
+ */
+const decoupleGraphPath = (nodePath: Array<HoisterWorkTree>): HoisterWorkTree => {
+  let parentNode = nodePath[0];
+  for (const node of nodePath.slice(1))
```
[the remark this answers is not in the document]
Comment (larixer) [reply packages/yarnpkg-pnpify/sources/hoist.ts:159]: Good idea, thanks, yes, it is called a lot, will do it in the next PR

**Comment by bertho-zero** [comment]
@larixer I created a minimal repro of a bug introduced from this PR,, I reduced my project as much as possible:

https://github.com/bertho-zero/yarn-bug-repro/tree/bug

**Comment by larixer** [comment]
@bertho-zero Thanks! I will check and let you know

**Comment by larixer** [comment]
@bertho-zero Apparently this bug is old and not related to this PR. I have pushed a fix to the new PR, could you check please
https://github.com/yarnpkg/berry/pull/1453
`yarn set version from sources --branch 1453`

**Comment by bertho-zero** [comment]
I've been on Yarn 2 for several months and I only had this bug with this PR, or with master after the merge.

The problem seems to be fixed, thank you.

**Pull request merged**